cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+import io
+import zipfile
+from concurrent.futures import ThreadPoolExecutor
+from typing import Callable, Dict
+
+from lxml import etree as ET
+from simple_salesforce.api import Salesforce
+
+from cumulusci.core.dependencies.utils import TaskContext
+from cumulusci.salesforce_api.metadata import ApiRetrieveUnpackaged
+from cumulusci.salesforce_api.utils import get_simple_salesforce_connection
+
+
+class FileName:
+    folder_name: str
+    extension: str
+
+    def __init__(self, folder_name, extension):
+        self.folder_name = folder_name
+        self.extension = extension
+
+
+PROFILE_FILE = FileName("profiles/", ".profile")
+PERMISSIONSET_FILE = FileName("permissionsets/", ".permissionset")
+FILES_TO_BE_CLEANED = [PROFILE_FILE, PERMISSIONSET_FILE]
+
+
+class PermissionElementXPath:
+    permission_xpath: str
+    name_xpath: str
+
+    def __init__(self, permission_xpath, name_xpath):
+        self.permission_xpath = permission_xpath
+        self.name_xpath = name_xpath
+
+    def return_parent(self, element: ET.Element):
+        return element.find(self.name_xpath).text.split(".")[0]
+
+    def return_name(self, element: ET.Element):
+        return element.find(self.name_xpath).text
+
+
+OBJECT_PERMISSION = PermissionElementXPath(".//objectPermissions", "object")
+RECORDTYPE_PERMISSION = PermissionElementXPath(
+    ".//recordTypeVisibilities", "recordType"
+)
+FIELD_PERMISSION = PermissionElementXPath(".//fieldPermissions", "field")
+USER_PERMISSION = PermissionElementXPath(".//userPermissions", "name")
+TAB_PERMISSION = PermissionElementXPath(".//tabVisibilities", "tab")
+APP_PERMISSION = PermissionElementXPath(".//applicationVisibilities", "application")
+APEXCLASS_PERMISSION = PermissionElementXPath(".//classAccesses", "apexClass")
+APEXPAGE_PERMISSION = PermissionElementXPath(".//pageAccesses", "apexPage")
+FLOW_PERMISSION = PermissionElementXPath(".//flowAccesses", "flow")
+CUSTOMPERM_PERMISSION = PermissionElementXPath(".//customPermissions", "name")
+CUSTOMSETTING_PERMISSION = PermissionElementXPath(".//customSettingAccesses", "name")
+CUSTOMMETADATA_PERMISSION = PermissionElementXPath(
+    ".//customMetadataTypeAccesses", "name"
+)
+
+PACKAGEXML_DICT = {
```

Comment:
At first read, I don't understand these data structures. A comment might be helpful.

Reply:
Yep will put a comment there